node/network/availability-distribution/src/requester/mod.rs

```diff
@@ -91,50 +114,75 @@ impl Requester {
 		ctx: &mut Context,
 		runtime: &mut RuntimeInfo,
 		update: ActiveLeavesUpdate,
-	) -> super::Result<()>
+	) -> Result<()>
 	where
 		Context: SubsystemContext,
 	{
 		tracing::trace!(target: LOG_TARGET, ?update, "Update fetching heads");
 		let ActiveLeavesUpdate { activated, deactivated } = update;
-		// Stale leaves happen after a reversion - we don't want to re-run availability there.
-		let activated = activated.and_then(|h| match h.status {
-			LeafStatus::Stale => None,
-			LeafStatus::Fresh => Some(h),
-		});
 		// Order important! We need to handle activated, prior to deactivated, otherwise we might
 		// cancel still needed jobs.
-		self.start_requesting_chunks(ctx, runtime, activated.into_iter()).await?;
+		if let Some(activated) = activated {
+			// Stale leaves happen after a reversion - we don't want to re-run availability there.
+			if let LeafStatus::Fresh = activated.status {
+				self.start_requesting_chunks(ctx, runtime, activated).await?;
+			}
+		}
 		self.stop_requesting_chunks(deactivated.into_iter());
 		Ok(())
 	}
 
-	/// Start requesting chunks for newly imported heads.
+	/// Start requesting chunks for newly imported head.
+	///
+	/// This will also request [`SESSION_ANCESTRY_LEN`] leaf ancestors from the same session
+	/// and start requesting chunks for them too.
 	async fn start_requesting_chunks<Context>(
 		&mut self,
 		ctx: &mut Context,
 		runtime: &mut RuntimeInfo,
-		new_heads: impl Iterator<Item = ActivatedLeaf>,
-	) -> super::Result<()>
+		new_head: ActivatedLeaf,
+	) -> Result<()>
 	where
 		Context: SubsystemContext,
 	{
-		for ActivatedLeaf { hash: leaf, .. } in new_heads {
-			let cores = get_occupied_cores(ctx, leaf).await?;
+		let ActivatedLeaf { hash: leaf, .. } = new_head;
+		let ancestors_in_session =
+			get_block_ancestors_in_same_session(ctx, leaf, Self::LEAF_ANCESTRY_LEN_WITHIN_SESSION)
+				.await
+				.unwrap_or_else(|err| {
+					tracing::debug!(
+						target: LOG_TARGET,
+						leaf = ?leaf,
+						"Failed to fetch leaf ancestors in the same session due to an error: {}",
+						err
+					);
+					Vec::new()
+				});
+		for hash in std::iter::once(leaf).chain(ancestors_in_session.clone()) {
+			let cores = get_occupied_cores(ctx, hash).await?;
 			tracing::trace!(
 				target: LOG_TARGET,
 				occupied_cores = ?cores,
 				"Query occupied core"
 			);
-			self.add_cores(ctx, runtime, leaf, cores).await?;
+			self.add_cores(ctx, runtime, hash, cores).await?;
 		}
+
+		// Save leaf ancestors for tasks pruning in the future.
+		self.relay_ancestors.insert(leaf, ancestors_in_session);
+
 		Ok(())
 	}
 
 	/// Stop requesting chunks for obsolete heads.
 	///
 	fn stop_requesting_chunks(&mut self, obsolete_leaves: impl Iterator<Item = Hash>) {
-		let obsolete_leaves: HashSet<_> = obsolete_leaves.collect();
+		let obsolete_leaves: HashSet<_> = obsolete_leaves
```

Comment:
I don't see how this can work. If we always remove the ancestors as well, then this will have the same effect as not taking the ancestors into account at all. For each obsolete leaf we also remove all its ancestors - why add them in the first place?

Reply:
That's true, thanks
Should we instead schedule it? Like keep an LRU of fetch tasks for ancestors?

Reply:
If we only add the actually activated head to [`FetchTask::live_in`](https://github.com/paritytech/polkadot/blob/a6e1f37694b2b46b7132216181af4568017a23ca/node/network/availability-distribution/src/requester/fetch_task/mod.rs#L67), even if it was actually spawned for an ancestor - then we should not need to keep track of ancestors separately. (We should change docs accordingly)

For lengthening the fetch time a bit, the cleanup could become a two phase process:

- Before updating `FetchTask::live_in` for the tasks, we find those that already have an empty `live_in` set. We remove it if the block number of the currently deactivated leaf differed from the block number of the last deactivated leaf.*)
- Then we remove any deactivated heads in `FetchTask::live_in`, but don't yet remove the task.


We can keep track of block numbers for all active leaves and always store the block number of the last deactivated.

This should lengthen the fetching phase for another slot (most of the time) and should be robust even in the event of chain reversions.

Note: While we will only use activated heads in `live_in` we do need to use the correct relay parent  for [`relay_parent`](https://github.com/paritytech/polkadot/blob/a6e1f37694b2b46b7132216181af4568017a23ca/node/network/availability-distribution/src/requester/fetch_task/mod.rs#L121)

*) That's a simple heuristic to make sure we are really lengthening the time - it is not 100% but very close and keeps the code a little simpler (especially with chain reversions).